packages/prio3/src/prio3.ts

```diff
@@ -27,70 +27,93 @@ interface DecodedShare {
   inputShare: bigint[];
   proofShare: bigint[];
   blind: Buffer | null;
-  hint: Buffer | null;
+  jointRandPartsHint: Buffer[] | null;
 }
 
 interface Share {
   blind: Buffer;
-  hint: Buffer;
+  jointRandPart: Uint8Array;
   inputShare: bigint[];
   proofShare: bigint[];
   inputShareSeed?: Buffer;
   proofShareSeed?: Buffer;
+  jointRandPartsHint?: Uint8Array[];
 }
 
-const DOMAIN_SEPARATION_TAG = Buffer.from(`${VDAF_VERSION} prio3`, "ascii");
+const DOMAIN_SEPARATION_TAG_BASE = Buffer.from(VDAF_VERSION, "ascii");
+const DST_LEN = DOMAIN_SEPARATION_TAG_BASE.length + 4;
 export class Prio3<Measurement> implements Prio3Vdaf<Measurement> {
   readonly rounds = 1;
   readonly verifyKeySize: number;
 
   constructor(
     public readonly prg: PrgConstructor,
     public readonly flp: Flp<Measurement>,
-    public readonly shares: number
+    public readonly shares: number,
+    public readonly algorithmId: number
   ) {
     this.verifyKeySize = prg.seedSize;
   }
 
+  private domainSeparationTag(): Buffer {
+    const buffer = Buffer.alloc(DST_LEN);
+    DOMAIN_SEPARATION_TAG_BASE.copy(buffer, 0);
+    buffer.writeUInt32BE(this.algorithmId, DOMAIN_SEPARATION_TAG_BASE.length);
+    return buffer;
+  }
+
   private async pseudorandom(
     len: number,
     seed?: Buffer,
-    info?: number
+    info?: number | Buffer
   ): Promise<bigint[]> {
     const { prg, field } = this;
-    return prg.expandIntoVec(
-      field,
-      seed || prg.randomSeed(),
-      typeof info === "number"
-        ? Buffer.from([...DOMAIN_SEPARATION_TAG, info])
-        : DOMAIN_SEPARATION_TAG,
-      len
-    );
+
+    const domainSeparationTag = this.domainSeparationTag();
+
+    let prgInfo;
+    if (typeof info === "number") {
+      prgInfo = Buffer.from([...domainSeparationTag, info]);
+    } else if (info instanceof Buffer) {
+      prgInfo = Buffer.concat([domainSeparationTag, info]);
```

Comment:
Maybe this is a TypeScript thing I don't understand but why do we use `typeof ===` in one spot and `instanceof` in the other?